### PR TITLE
refactor: Pass `this` pointer as first param to s11n callbacks.

### PR DESCRIPTION
**Pull request by iphydf**
This change is [Reviewable](https://reviewable.io/reviews/TokTok/c-toxcore/2575)

### CODE REVIEW

File: toxcore/bin_pack.h [hunk 1]
```diff
@@ -16,6 +16,23 @@ extern "C" {
 
 /**
  * @brief Binary serialisation object.
+ *
+ * Some notes on parameter order:
+ *
+ * - We pass the `obj` pointer as `this`-like pointer first to the callbacks.
```
Comment (nurupo) [inline toxcore/bin_pack.h:22]: Isn't `Bin_Pack *bp` the `this` of `bin_[un]pack_()` functions, not `void *obj`?

File: toxcore/bin_pack.h [hunk 2]
```diff
@@ -34,7 +51,7 @@ typedef bool bin_pack_cb(Bin_Pack *bp, const Logger *logger, const void *obj);
  * @param arr is the object array as void pointer.
  * @param index is the index in the object array that is currently being packed.
  */
-typedef bool bin_pack_array_cb(Bin_Pack *bp, const Logger *logger, const void *arr, uint32_t index);
+typedef bool bin_pack_array_cb(const void *arr, uint32_t index, const Logger *logger, Bin_Pack *bp);
```
Comment (nurupo) [inline toxcore/bin_pack.h:54]: Re-arrange `@param` in function doc too?

File: toxcore/bin_pack.h [hunk 3]
```diff
@@ -46,7 +63,7 @@ typedef bool bin_pack_array_cb(Bin_Pack *bp, const Logger *logger, const void *a
  * @retval UINT32_MAX in case of errors such as buffer overflow.
  */
 non_null(1) nullable(2, 3)
-uint32_t bin_pack_obj_size(bin_pack_cb *callback, const Logger *logger, const void *obj);
+uint32_t bin_pack_obj_size(bin_pack_cb *callback, const void *obj, const Logger *logger);
```
Comment (nurupo) [inline toxcore/bin_pack.h:66]: Re-arrange `@param` in function doc too?

File: toxcore/bin_pack.h [hunk 4]
```diff
@@ -65,23 +82,23 @@ uint32_t bin_pack_obj_size(bin_pack_cb *callback, const Logger *logger, const vo
  * @retval false if an error occurred (e.g. buffer overflow).
  */
 non_null(1, 4) nullable(2, 3)
-bool bin_pack_obj(bin_pack_cb *callback, const Logger *logger, const void *obj, uint8_t *buf, uint32_t buf_size);
+bool bin_pack_obj(bin_pack_cb *callback, const void *obj, const Logger *logger, uint8_t *buf, uint32_t buf_size);
```
Comment (nurupo) [inline toxcore/bin_pack.h:85]: Re-arrange `@param` in function doc too?

File: toxcore/tox_events.c [hunk 5]
```diff
@@ -102,8 +102,8 @@ Tox_Events *tox_events_iterate(Tox *tox, bool fail_hard, Tox_Err_Events_Iterate
     return state.events;
 }
 
-non_null(1) nullable(2, 3)
-static bool tox_events_pack(Bin_Pack *bp, const Logger *logger, const void *obj)
+non_null(2) nullable(1, 3)
+static bool tox_events_pack(const void *obj, const Logger *logger, Bin_Pack *bp)
```
Comment (nurupo) [inline toxcore/tox_events.c:106]: Those `non_null()` and `nullable()` don't match the behavior before the change. Should be `non_null(3) nullable (1, 2)` to match the previous behavior. Was that intentional?

**Review by nurupo — CHANGES_REQUESTED** [review]

**Review by Green-Sky — APPROVED** [review]

File: toxcore/bin_pack.h (hunk 3 above)
Comment (iphydf) [reply toxcore/bin_pack.h:64]: Done.

File: toxcore/bin_pack.h (hunk 4 above)
Comment (iphydf) [reply toxcore/bin_pack.h:85]: Done.

File: toxcore/tox_events.c (hunk 5 above)
Comment (iphydf) [reply toxcore/tox_events.c:106]: Yes, luckily cimple warns about it. Fixed.

Comment (nurupo) [reply toxcore/tox_events.c:106]: I don't see it being fixed, looks like nothing has changed. Did you forget to stage the hunk or something?

File: toxcore/bin_pack.h (hunk 4 above)
Comment (nurupo) [reply toxcore/bin_pack.h:85]: Looks good.

File: toxcore/bin_pack.h (hunk 3 above)
Comment (nurupo) [reply toxcore/bin_pack.h:64]: I don't see it being fixed, looks like nothing has changed. Did you forget to stage the hunk or something?

File: toxcore/bin_pack.h (hunk 2 above)
Comment (iphydf) [reply toxcore/bin_pack.h:52]: No, I don't think so. There doesn't seem to be anything in the function doc that needs to be rearranged here.

File: toxcore/bin_pack.h (hunk 3 above)
Comment (iphydf) [reply toxcore/bin_pack.h:64]: Fixed now :).

File: toxcore/tox_events.c (hunk 5 above)
Comment (iphydf) [reply toxcore/tox_events.c:106]: Fixed now.

**Review by nurupo — APPROVED** [review]

**Pull request merged**